### PR TITLE
Mixin for plotting context vectors

### PR DESCRIPTION
**Pull request by karangrewal**
This new mixin is similar to `PlotHiddenActivations`, but plots context vectors (generated by a dendritic network directly from the input) instead. [Here](https://drive.google.com/file/d/10oHOqLZWsRYO5PYdZcdA_WNEqDCaoAnP/view?usp=sharing) is the resulting plot, where cell c, k gives the fraction of instances of category c for which feature k in the context vector was 'on'. It requires a new plotting function for visualization.

The easiest use case is to set `include_names` or `include_modules` in the config to refer to the modules that output the context vector, then a hook is registered that tracks its outputs.

In the future, we'll have to address how to unify all these mixins, since
- many of them have related functions, and we can probably reduce the amount of code that achieves the same goal, and
- the memory costs grow with the number of such mixins used in an experiment (e.g., each such mixin tracks target variables (`context_targets` in this case), which are all just duplicates of each other,

but I wouldn't worry about this for now.

### CODE REVIEW

**Review by mvacaporale — APPROVED** [review]
Looks good. I do agree that we keep reusing a lot of code. That's not too bad, but it's definitely building technical debt. To summarize what you said and add to it:
* There's a potential memory issue of saving the same tensors repeatedly; this could be unlikely since we probably only run with a few metric-mixins at a time.
* There's of course a performance issue of repeating the same functionality to log these metrics.
* There's an inheritance issue where the attribute names can conflict with one another. But we skirt around this by using different names for the same thing, like calling the hook manager `self.context_hook`. This is slightly awkward but it works.
* There's also the difficulty of understanding which mixin does what and so the readability of our codebase gets diminished. 
* There's also the missed opportunity of easy-to-implement metrics. We could have one mixin to do all these metrics. The user would simply define a hook and associated functions for plotting and then pass these through the config.

If you're interested in making such a mixin (which I do encourage if another PR like this comes up), then please let me know. I'd be happy to brainstorm with you and discuss implementation if you'd like as I've already put some thought into it.

File: nupic/research/frameworks/dendrites/metrics/metrics.py [hunk 1]
```diff
@@ -199,6 +199,48 @@ def hidden_activations_by_unit(activations, targets):
...
+
+        device = contexts.device
+
+        # Assume the following:
+        # - target values are zero-based
+        # - the largest target value in the batch is that amongst all data
+        num_categories = 1 + targets.max().item()
+        _, dim_context = contexts.size()
+
+        _contexts_by_class = torch.zeros((0, dim_context))
```
[the remark this answers is not in the document]
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/metrics/metrics.py:225]: My only reason is because the function has the same name (i.e., `context_by_class`), so I wanted to distinguish it. I guess I could just call it `cbc` or something of that sort.

[the remark this answers is not in the document]
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/metrics/metrics.py]: Fixed.

**Comment by karangrewal** [comment]
That's a good summary. I'll keep you in the loop as more mixins like this come up, and we can figure out how to address this replication issue. It's not top priority for me right now, so that's why I'm not acting on it.
